CDS/scripts/atom_image.pl

```diff
@@ -30,12 +30,21 @@ sub is_imageurl_valid {
 print "INFO: Attempting to get image URL from CAPI\n";
 
 my $ua = LWP::UserAgent->new;
-my $req = $ua->request(GET 'https://internal.content.guardianapis.com/atom/media/'.$store->substitute_string($ENV{'atom_id'}));
+my $response = $ua->request(GET 'https://internal.content.guardianapis.com/atom/media/'.$store->substitute_string($ENV{'atom_id'}));
 
-my $capi = decode_json($req->content);
+if ($response->code == 503) {
```

Comment:
This needs to be structured as a loop with a timeout, because if you retry immediately after getting a 503 you'll probably get another.

Also, the use of `my` in the line below limits the scoping of `$response` below to only the if {} block which is not what you want.

You need something like this:

```
my $response;
while(true) {
  $response = $ua->request(....);
  if($response->code >=400 || response->code <=499){
     print $response->content;
     print "\n-ERROR: CAPI returned " . $response->code . ", actual error is logged above.\n";
     exit(1);
  } elif ($response->code >=500 || $response->code <= 599){
     print $response->content;
     print "\n-ERROR: CAPI returned " . $response->code . ", actual error is logged above.\n";
     exit(1) if($retries>=$max_retries);
     sleep($delay);
     continue;
  } elif($response->code==200){
      print "*INFO: Got response from CAPI\n";
      last;
  } else {
      print "-ERROR: Unexpected status code ".$response->code." from CAPI, retrying";
      sleep($delay);
      continue
}
```

That way client-side errors (40x) are fatal since retrying them should give the same result. Server-side errors (50x) are retryable in the hope that the server sorts itself out, 200 is success and anything else is unexpected and should be retried (assuming that it is a server weirdness).

Plus, we get a useful log trace; if CAPI is misbehaving for some reason it makes it much simpler to go and ask the CAPI team for help.

Reply:
Done